OA-PASS/java-fedora-client
# Maven central compatibility

## Overview

* Adds the javadoc plugin, for attaching javadoc artifacts when deploying (a maven central requirement)
  * Lots of javadoc fixes
* Adds the release and GPG plugins, for releasing and signing artifacts
* Adds license, developers, and scm sections to the POM, as required by Maven Central

With this, the release plugin should work, and maven artifacts should pass maven central validation

Patches derived from this will be applied to `maintenance-0.3` in support of #60 

## To test
Use the release plugin to create a release locally

    export RELEASE=0.4.1
    export NEXT=0.4.2-SNAPSHOT

    mvn release:prepare -DreleaseVersion=$RELEASE -Dtag=$RELEASE -DdevelopmentVersion=$NEXT -DautoVersionSubmodules=true -DpushChanges=false

The next step will upload to the maven central staging (would be nice to try out.  There is no requirement that it be promoted to maven central as a true release if you do that; it can be cancelled and deleted from stage) .  If you want to do this, [sign up with sonatype](https://issues.sonatype.org/secure/Signup!default.jspa) and ping @birkland to create a JIRA ticket with the Sonatype people to grant access to sonatype for deploying DC artifacts to the OSS snapshot repo and maven central.  You'd also want to generate [some gpg artifact signing keys](https://central.sonatype.org/pages/working-with-pgp-signatures.html) 

    mvn release:perform -Dgoals=deploy -DlocalCheckout

If you get that far and actualy did upload to the maven central staging repo, you can validate the uploaded artifacts by doing the following:
* Go to https://oss.sonatype.org/index.html  and log in
* Click Staging Repositories in left navigation, under "Build Promotion"
* Either search for "dataconservancy" in upper right search box, or scroll through the liset
* Select repository and verify that artifacts present in the Content tab
* Read https://central.sonatype.org/pages/releasing-the-deployment.html
* Click `close` near the TOP of the page.  This will make it undergo validation, which takes a little time
* Click `refresh` near the TOP of the page, which will reset the status.  If the operation has completed successfully, you'll see the repository in a "closed" state.  Feel free to look at the activity tab to see the validation steps.
* Click `release` to release (I don't think you want to to this when testing a PR!), or `drop` to drop

If this were a real release, you'd then push the tags created by the maven release plugin to github like `git push origin --tag`, and also push the branch you're on (e.g. `master`).  But since this is a PR, you'd do `git tag -d $RELEASE` to delete the tag, and also a `git reset --hard` to walk back your local tree.  The release plugin will have made commits bumping the version to the next snap.
